src/com/amazon/ion/impl/IonReaderLookaheadBuffer.java

```diff
@@ -0,0 +1,872 @@
+        // This is the start of a new top-level value. Mark the write index in case this value ends up exceeding
+        // the maximum size and needs to be truncated.
+        valueStartAvailable = pipe.available();
+        startNewValue();
+    }
+
+    /**
+     * Constructs a wrapper with the given configuration.
+     * @param configuration the configuration for the new instance.
+     * @param inputStream an InputStream over binary Ion data.
+     */
+    public IonReaderLookaheadBuffer(final IonBufferConfiguration configuration, final InputStream inputStream) {
+        super(configuration, inputStream);
+        pipe.registerNotificationConsumer(
+            new ResizingPipedInputStream.NotificationConsumer() {
+                @Override
+                public void bytesConsolidatedToStartOfBuffer(int leftShiftAmount) {
+                    // The existing data in the buffer has been shifted to the start. Adjust the saved indexes
+                    // accordingly. -1 indicates that all indices starting at 0 will be shifted.
+                    shiftIndicesLeft(-1, leftShiftAmount);
+                }
+            }
+        );
+        pageSize = configuration.getInitialBufferSize();
+        inProgressVarUInt = new VarUInt();
+        annotationSids = new ArrayList<Integer>(3);
+        reset();
+    }
+
+    /**
+     * Resets the `inProgressVarUInt`.
+     * @param location the VarUInt's location.
+     */
+    private void initializeVarUInt(final VarUInt.Location location) {
+        inProgressVarUInt.reset(location);
+        state = State.READING_HEADER;
+    }
+
+    /**
+     * Reads one byte, if possible.
+     * @return the byte, or -1 if none was available.
+     * @throws Exception if thrown by a handler method or if an IOException is thrown by the underlying InputStream.
+     */
+    private int readByte() throws Exception {
+        if (pipe.availableBeyondBoundary() == 0 && fillPage(1) < 1) {
+            return -1;
+        }
+        int b;
+        if (isSkippingCurrentValue()) {
+            // If the value is being skipped, the byte will not have been buffered.
+            b = getInput().read();
+        } else {
+            b = pipe.peek(peekIndex);
+            pipe.extendBoundary(1);
+            peekIndex++;
+        }
+        return b;
+    }
+
+    /**
+     * Reads a VarUInt. NOTE: the VarUInt must fit in a `long`. This is not a true limitation, as IonJava requires
+     * VarUInts to fit in an `int`.
+     * @throws Exception if thrown by a handler method or if an IOException is thrown by the underlying InputStream.
+     */
+    private void readVarUInt() throws Exception {
+        int currentByte;
+        while (inProgressVarUInt.numberOfBytesRead < MAXIMUM_SUPPORTED_VAR_UINT_BYTES) {
+            currentByte = readByte();
+            if (currentByte < 0) {
+                return;
+            }
+            inProgressVarUInt.numberOfBytesRead++;
+            eventHandler.onData(1);
```

Comment:
This is correct, but I wonder if it would be more useful for a user to be notified once of a `VarUInt`'s worth of data. If the callback isn't trivial then calling it several times per value could add up.